### PR TITLE
Adding an event bus, #422

### LINKED ISSUE #422: Users can be notified when the indexer ingests documents/transactions
(original message below)

We'd like users to be able to receive notifications from the indexer (and other areas, in the future) so that modules can react to transactions flowing through the Crux node. Potential use cases to consider:
* Collecting metrics (#384)
* Subscribing to indexed transactions to update downstream consumers
* Re-publish the events to Kafka topics.

Potential solution - event bus:
* must be very light weight. it shouldn't be slow, but also not complex
* it should publish the events async, potentially in a thread per subscriber module (not subscription).
* events may be split by topics, or the filtering could simply be done by the listener. 
* the events need to adhere to specs - a few keys might be mandatory, like `crux.event-bus/topic` or similar.

----
A “hook” for processed txes would ease both user subscriptions to db updates and our work with #161 . It can be a hook, an event or a signal. In its most minimal form it's just a callback accepting tx meta `#:crux.tx{:tx-id 1, :tx-time 1}`, so users could leverage that when working with tx-log. 

Perhaps, an internal hook for accepted non-processed transactions could also ease development of new tx-log consumers, maybe even streamline the existing ones; but here I'm not so sure.

This issue is here to accumulate more thought and feedback from the core devs and the community.

### PR DESCRIPTION
**Pull request by jarohen**
(fixes #422)

raising as a draft PR for early feedback.

* crux.event-bus which keeps track of who's listening to what, filtered by event-type
* added an initial set of events to the indexer, for notifications about indexing docs and txs

TODO:
* [x] doc-strings/javadocs
* [x] spec-asserting the events
* [x] do we need more events for an initial pass? I'm leaning towards 'no for now', can be driven by ops monitoring use cases
* [ ] how should this be exposed in the public API?

### CODE REVIEW

File: crux-core/test/crux/event_bus_test.clj [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+      (sut/listen bus {::sut/event-types #{:foo}} #(swap! !filtered-events conj %))
+      (sut/listen bus #(swap! !unfiltered-events conj %))
+
+      (sut/send bus {::sut/event-type :foo, :value 2})
+      (sut/send bus {::sut/event-type :bar, :value 1})
+
+      ;; just to ensure all the jobs are handled
+      ;; - we don't guarantee this if the node is shut down
+      (Thread/sleep 100))
```
Comment (hraberg) [inline crux-core/test/crux/event_bus_test.clj]: I know this should normally work, but could we get some promise and a deref with timeout here so we can remove the sleep?

**Review by hraberg — APPROVED** [review]

File: crux-core/test/crux/event_bus_test.clj (hunk 1 above)
Comment (jarohen) [reply crux-core/test/crux/event_bus_test.clj]: ah, yes, of course - doh.

**Comment by jonpither** [comment]
Interested to see how metrics will make use of this, i.e. timing query runs. Looks good to me, was mildly wondering for aesthetics if `crux.event-bus` should be `crux.bus`.

File: crux-core/src/crux/tx.clj [hunk 2]
```diff
@@ -340,17 +350,20 @@
                              (->> (vals docs-to-remove)
                                   (map #(idx/doc-predicate-stats % true))))]
 
+      (db/put-objects object-store docs)
+
```
Comment (jonpither) [inline crux-core/src/crux/tx.clj:354]: Wearing my pedantic hat, possibly wouldn't have this blank line:-)

**Review by jonpither — APPROVED** [review]

**Pull request merged**